### ISSUE TITLE
Seeking feedback on a plug-in-like system for environment discovery

### ISSUE THREAD

**Issue by brettcannon** [python-poetry/poetry#6014]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Issue

I asked in https://github.com/python-poetry/poetry/discussions/5938 for feedback on a potential design to let tools like Poetry to participate in telling tools like VS Code and the Python Launcher for Unix where the virtual environments they manage exist, as well has help in finding Python interpreters. But I realized that directly opening a discussion may not be the workflow you prefer, hence this issue.

**Comment by neersighted** [comment]
This sounds interesting, but I think would require a breaking change in Poetry. Is the idea to have a standard way to interrogate the Poetry binary (or module) about which environments it is responsible for, and which directory prefixes they are associated with?

**Comment by brettcannon** [comment]
> Is the idea to have a standard way to interrogate the Poetry binary (or module) about which environments it is responsible for

Correct. Right now every tool that manages environments don't have a standardized way to share what environments they have for a certain directory/workspace. This idea is trying to solve that somehow, and in a way that allows users to write the locator if the tool in question hasn't done it or isn't interested.

**Comment by neersighted** [comment]
Also, I think an entrypoint would be the best, most portable way to discover this functionality and call it (with some standardized Protocol describing the function-level interface).

**Comment by brettcannon** [comment]
> Are you interested in asking a tool what venv(s) it knows of for a given path, or asking it what venvs it is aware of period

The latter. Listing every single environment is not necessarily useful to know as that most likely doesn't apply to the current workspace. But if a tool wanted to list every environment as applicable to a workspace they could do that.

> I guess I'm curious what data model you envision.

It's in a link in the Discussion a linked to, but for ease of access it's https://github.com/brettcannon/python-launcher/discussions/168 .


> I think an entrypoint would be the best, most portable way to discover this functionality and call it (with some standardized Protocol describing the function-level interface).

I'm trying not to tie this to Python as this could be used by editors who may not be written in Python. Plus, it requires finding the installation of a tool to then find its `entry_points.txt` file to find out what entry points it defines which may or may not exist (e.g. PyFlow is written entirely in Rust).

**Comment by neersighted** [comment]
It sounds like you were interested in the former of my two options, actually -- some sort of exhaustive mutual mapping that contains all workspaces and environments. It's possible, but only with changes in Poetry as currently we only map workspace path -> environment.

I'm fine with the approach you've outlined with a binary that returns a JSON blob. Do you have any ideas what input parameters might look like?

**Comment by brettcannon** [comment]
> Do you have any ideas what input parameters might look like?

I have outlined some possibilities in https://github.com/brettcannon/python-launcher/discussions/168#discussioncomment-2267089 .
